### PR TITLE
ceph: compare ceph commit ID for upgrade check

### LINKED ISSUE #7964: Child controllers should reconcile if the ceph image changed
**Is this a bug report or feature request?**
* Bug Report

**Deviation from expected behavior:**
Currently the reconcile for child controllers such as file and object are only triggered if the Ceph version changed. 

The version check is currently only based on the build number.

In the repro seen where mds and rgw were not updated, the two versions before/after the upgrade were:
```
ceph version 14.2.11-139.0.hotfix.xxxxxxx.el8cp
(5c0dc966af809fd1d429ec7bac48962a746af243) nautilus (stable)": 7,
"ceph version 14.2.11-139.el8cp (b8e1f91c99491fb2e5ede748a1c0738ed158d0f5)
nautilus (stable)": 4
```

The build numbers 14.2.11-139 are all equivalent. The rest of the build version is ignored by Rook for the comparison.

**Expected behavior:**
The child controllers should all reconcile if the Ceph image changes, not just if the ceph version changed inside the image. The update to the image may not always include an update to the version number, at least in some downstream hotfix builds. 

**How to reproduce it (minimal and precise):**

For a simple repro:
- Deploy Rook with a ceph image such as `ceph/ceph:v16.2.4`
- Create object.yaml and filesystem.yaml to ensure child controllers are active
- Change the docker tag locally to simulate an image change: `docker tag ceph/ceph:v16.2.4 ceph/ceph:v16.2.4-alternate`
- Apply the new image to the CephCluster CR: `ceph/ceph:v16.2.4-alternate`
- Notice that all the daemons are updated except rgw and mds

### PR DESCRIPTION
**Pull request by sp98**
.Compare Image ID while checking for upgrades. This will be useful in the case where ceph version is same but  image ID is different and also when build number has not changed for a hotfix build. 

For example:
**existing ceph version** - 
```ceph version 14.2.11-139.el8cp (b8e1f91c99491fb2e5ede748a1c0738ed158d0f5) nautilus (stable) ```

**new ceph version with hotfix** - 
```ceph version 14.2.11-139.0.hotfix.bz1959254.el8cp (5c0dc966af809fd1d429ec7bac48962a746af243) nautilus (stable)```

Signed-off-by: Santosh Pillai <sapillai@redhat.com>

**Description of your changes:**

**Which issue is resolved by this Pull Request:**
Resolves #7964

Commit ID Regex test: https://play.golang.org/p/S5wBQrueC_m
**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/master/development-flow.html#commit-structure).
- [ ] **Skip Tests for Docs**: Add the flag for skipping the build if this is only a documentation change. See [here](https://github.com/rook/rook/blob/master/INSTALL.md#skip-ci) for the flag.
- [ ] **Skip Unrelated Tests**: Add a flag to run tests for a specific storage provider. See [test options](https://github.com/rook/rook/blob/master/INSTALL.md#test-storage-provider).
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/master/development-flow.html#submitting-a-pull-request)
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.
- [ ] Pending release notes updated with breaking and/or notable changes, if necessary.
- [ ] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [ ] Code generation (`make codegen`) has been run to update object specifications, if necessary.

### CODE REVIEW

File: pkg/operator/ceph/version/version.go [hunk 1]
```diff
@@ -257,7 +269,7 @@ func IsSuperior(a, b CephVersion) bool {
 	if a.Major == b.Major {
 		if a.Minor == b.Minor {
 			if a.Extra == b.Extra {
-				if a.Build > b.Build {
+				if a.Build > b.Build || (a.Build == b.Build && a.BuildSuffix != b.BuildSuffix) {
```
Comment (travisn) [inline pkg/operator/ceph/version/version.go:272]: same functionality, but to keep the structure similar to other checks:
```suggestion
				if a.Build > b.Build {
					return true
				}
				if a.Build == b.Build {
					if a.BuildSuffix != b.BuildSuffix) {
						return true
					}
				}			
```

File: pkg/operator/ceph/version/version.go [hunk 2]
```diff
@@ -128,7 +132,13 @@ func ExtractCephVersion(src string) (*CephVersion, error) {
 		}
 	}
 
-	return &CephVersion{major, minor, extra, build}, nil
+	// get build suffix if running on a build release
+	mmm := buildSuffixPattern.FindStringSubmatch(src)
```
Comment (travisn) [inline pkg/operator/ceph/version/version.go:136]: nit: how about a more descriptive name than `mmm`? It's interesting to follow the pattern of the rest of the method, but I'd vote for descriptive names instead of a pattern like that. :)

**Review by travisn — CHANGES_REQUESTED** [review]
I see this fixes the version check issue, but I think there is a bigger issue as well... If the `cephVersion.image` field changes for any reason, we should make sure all the ceph daemons get the new image, even if the ceph version is unchanged. Otherwise, we could have daemons running different images until the operator is restarted or there is some other update to the cluster CR that causes a full reconcile. This is a corner case and could be a separate PR, but I'd still propose we address it.

File: pkg/operator/ceph/version/version.go (hunk 2 above)
Comment (sp98) [reply pkg/operator/ceph/version/version.go]: updated. Using `<>Match` as variable names

**Comment by sp98** [comment]
> I see this fixes the version check issue, but I think there is a bigger issue as well... If the `cephVersion.image` field changes for any reason, we should make sure all the ceph daemons get the new image, even if the ceph version is unchanged. Otherwise, we could have daemons running different images until the operator is restarted or there is some other update to the cluster CR that causes a full reconcile. This is a corner case and could be a separate PR, but I'd still propose we address it.

Added another commit that compares image ID. Now we check for both Build Suffix and Image ID.  I think we can remove Build suffix check and only keep image ID check as that would suffice for our use case.

File: pkg/operator/ceph/version/version.go [hunk 3]
```diff
@@ -27,10 +27,12 @@ import (
...
-	Major int
-	Minor int
-	Extra int
-	Build int
+	Major       int
+	Minor       int
+	Extra       int
+	Build       int
+	BuildSuffix string
+	Image       string
```
Comment (leseb) [inline pkg/operator/ceph/version/version.go:35]:
```suggestion
	CommitID       string
```

File: pkg/operator/ceph/cluster/version_test.go [hunk 4]
```diff
@@ -109,6 +110,57 @@ func TestDiffImageSpecAndClusterRunningVersion(t *testing.T) {
...
 	assert.NoError(t, err)
 	assert.False(t, m)
+
+	// 6 test - spec version and running cluster have same build but different build suffix --> we upgrade
+	fakeImageVersion = cephver.CephVersion{Major: 14, Minor: 2, Extra: 11, Build: 139,
+		BuildSuffix: "el8cp", Image: "5c0dc966af809fd1d429ec7bac48962a746af243"}
+	fakeRunningVersions = []byte(`
+		{
+			"overall": {
+				"ceph version 14.2.11-139.0.hotfix.bz1959254.el8cp (5c0dc966af809fd1d429ec7bac48962a746af243) nautilus (stable)": 2
```
Comment (leseb) [inline pkg/operator/ceph/cluster/version_test.go:120]: So this `139.0.hotfix.bz1959254` is the `Build`? and `el8cp` the `BuildSuffix`?

File: pkg/operator/ceph/version/version.go [hunk 5]
```diff
@@ -98,37 +104,50 @@ func (v *CephVersion) ReleaseName() string {
...
 	}
 
-	return &CephVersion{major, minor, extra, build}, nil
+	// get build suffix if running on a build release
+	buildSuffixMatch := buildSuffixPattern.FindStringSubmatch(src)
+	if buildSuffixMatch != nil {
+		buildSuffix = buildSuffixMatch[5]
+	}
+
+	imageMatch := imagePattern.FindStringSubmatch(src)
```
Comment (leseb) [inline pkg/operator/ceph/version/version.go:145]: Why is this needed?

File: pkg/operator/ceph/version/version.go [hunk 6]
```diff
@@ -65,6 +67,10 @@ var (
 	// So we need to detect the build version change
 	buildVersionPattern = regexp.MustCompile(`ceph version (\d+)\.(\d+)\.(\d+)\-(\d+)`)
 
+	buildSuffixPattern = regexp.MustCompile(`ceph version (\d+)\.(\d+)\.(\d+)\-(\d+).(\S+)`)
+
+	imagePattern = regexp.MustCompile(`\(([^)]+)\)`)
```
Comment (travisn) [inline pkg/operator/ceph/version/version.go:72]: Does this match everything from the first character to the first whitespace? Or what will match? Can you add a comment about it? I like this idea of getting the entire version string and not worry about the fields. Then I don't think we need to get the buildSuffix, we can just use this full image name and if the ceph build version comes up with even more unanticipated fields we will already detect it instead of needing more fields.

File: pkg/operator/ceph/version/version.go [hunk 7]
```diff
@@ -260,6 +283,14 @@ func IsSuperior(a, b CephVersion) bool {
 				if a.Build > b.Build {
 					return true
 				}
+				if a.Build == b.Build {
+					if a.BuildSuffix != b.BuildSuffix {
+						return true
+					}
+				}
+				if a.Image != b.Image {
```
Comment (travisn) [inline pkg/operator/ceph/version/version.go:291]: If the image changed we don't know if it's superior, right? We only know it's different. For superiority I think we don't need to know at a depth past `a.Build` so we could revert the changes to this method.

**Review by travisn — CHANGES_REQUESTED** [review]

File: pkg/operator/ceph/cluster/version_test.go (hunk 4 above)
Comment (sp98) [reply pkg/operator/ceph/cluster/version_test.go:120]: `139` is the build and anything after that is the build suffix `0.hotfix.bz1959254.el8cp` in this case.  The implementation in `master` branch only checks till `14.2.11-139`  .  [Reference](https://github.com/rook/rook/blob/52a9dfc26d65dd7b355eb23f2707c3aa65d20a62/pkg/operator/ceph/version/version.go#L66)

File: pkg/operator/ceph/version/version.go (hunk 5 above)
Comment (sp98) [reply pkg/operator/ceph/version/version.go:145]: trying to fetch the commit ID of the image here. I believe this should address the [commit](https://github.com/rook/rook/pull/8060#pullrequestreview-677678759)

File: pkg/operator/ceph/version/version.go (hunk 6 above)
Comment (sp98) [reply pkg/operator/ceph/version/version.go:72]: This would only return the commit ID. 
For example, it will return `b8e1f91c99491fb2e5ede748a1c0738ed158d0f5` from the source `ceph version 14.2.11-139.el8cp (b8e1f91c99491fb2e5ede748a1c0738ed158d0f5) nautilus (stable)`.  I was trying to address the [comment](https://github.com/rook/rook/pull/8060#pullrequestreview-677678759) with this comparison.

Comment (travisn) [reply pkg/operator/ceph/version/version.go:72]: Ah that makes sense. That image hash will be a good representation of whether anything changed. So if we get this image and it changed, we would trigger the child controllers, it should fully address the issue. There is no need to extract the buildSuffix now, right?

Comment (sp98) [reply pkg/operator/ceph/version/version.go:72]: Yes, checking commit hash should suffice and we can ignore the `buildSuffix`. This should fix the case when build number didn't change due to a hotfix.

File: pkg/operator/ceph/version/version.go [hunk 8]
```diff
@@ -65,6 +66,11 @@ var (
 	// So we need to detect the build version change
 	buildVersionPattern = regexp.MustCompile(`ceph version (\d+)\.(\d+)\.(\d+)\-(\d+)`)
```
Comment (leseb) [inline pkg/operator/ceph/version/version.go:67]: Why not include the `commitIDPattern` here? The commit is always present.

**Review by leseb — CHANGES_REQUESTED** [review]

File: pkg/operator/ceph/version/version.go (hunk 8 above)
Comment (sp98) [reply pkg/operator/ceph/version/version.go:67]: We can include it in the `VersionPattern` above  and not in the `buildVersionPattern`. The build number represented by `\-(\d+)` is not always present.

Comment (leseb) [reply pkg/operator/ceph/version/version.go:67]: Right, the overall pattern so `VersionPattern` will be better.

Comment (sp98) [reply pkg/operator/ceph/version/version.go:67]: checked a bit more. I don't think I can update the existing `versionPattern` or `buildVersionPattern`.  This is because the `ExtractCephVersion` function is also used to extract the ceph version from `cephCluster` CR status (via `ExtractCephVersionFromLabel()` function.  And in the `cephCluster` `status.version` does not have any commit ID. 

It looks like:
```
....
 message: Cluster created successfully
  phase: Ready
  state: Created
  storage:
    deviceClasses:
    - name: hdd
  version:
    image: ceph/ceph:v16.2.4
    version: 16.2.4-0
```


So the regex match would fail if I update `versionPattern`.

**Review by travisn — APPROVED** [review]

File: pkg/operator/ceph/version/version.go [hunk 9]
```diff
@@ -228,7 +240,10 @@ func IsIdentical(a, b CephVersion) bool {
 		if a.Minor == b.Minor {
 			if a.Extra == b.Extra {
 				if a.Build == b.Build {
-					return true
+					if a.CommitID == b.CommitID {
+						return true
+					}
+
```
Comment (leseb) [inline pkg/operator/ceph/version/version.go:246]: nit newline

**Review by leseb — CHANGES_REQUESTED** [review]
small nit

**Review by leseb — APPROVED** [review]

**Comment by leseb** [comment]
@sp98 please open an issue to follow with what we discussed yesterday, thanks.

**Comment by sp98** [comment]
> @sp98 please open an issue to follow with what we discussed yesterday, thanks.

opened #8100
